File: materials/orders-csharp/Staff.cs

// =========================================================
// Файл: staff.go
// Описание: Система управления персоналом склада.
// =========================================================

namespace Orders;

interface IWarehouseWorker
{
    void ProcessOrder();
    void AttendMeeting();
    void GetRest();
    void SwingingTheLead();
}

/// <summary>
/// HumanManager - Человек.
/// </summary>
class Human : IWarehouseWorker
{
    public void ProcessOrder()
    {
        Console.WriteLine("Manager is processing logic...");
    }

    public void AttendMeeting()
    {
        Console.WriteLine("Manager is boring at the meeting...");
    }

    public void GetRest()
    {
        Console.WriteLine("Manager is taking a break...");
    }

    public void SwingingTheLead()
    {
        Console.WriteLine("Manager is watching reels...");
    }
}


/// <summary>
/// RobotPacker - Робот.
/// </summary>
class RobotPacker : IWarehouseWorker
{
    public required string Model { get; init; }

    public void ProcessOrder() {
        Console.WriteLine("Robot " + Model + " is packing boxes...");
    }

    public void GetRest() {
        Console.WriteLine("Robot was taken for maintenance");
    }

    public void AttendMeeting() {
        Console.WriteLine("ERROR: Robot cannot attend meetings");
    }

    public void SwingingTheLead() {
        throw new InvalidOperationException("CRITICAL ERROR: Robot cannot waste our money (we hope so)");
    }
}

static class Warehouse
{
    /// <summary>
    /// ManageWarehouse - функция, которая работает со списком работников.
    /// </summary>
    /// <param name="workers"></param>
    public static void ManageWarehouse(IWarehouseWorker[] workers)
    {
        Console.WriteLine("\n--- Warehouse Shift Started ---");

        foreach (var worker in workers) {
    		worker.ProcessOrder();
    		worker.AttendMeeting();
    		worker.GetRest();
    		worker.SwingingTheLead();
    	}
    }   
}

File: materials/orders/rust/orders/src/staff.rs
//! Модуль `staff`.
//!
//! Описание: система управления персоналом склада.

pub trait WarehouseWorker {
    fn process_order(&self);
    fn attend_meeting(&self);
    fn get_rest(&self);
    fn swing_the_lead(&self);
}

/// [`HumanManager`] - человек
pub struct HumanManager;

impl WarehouseWorker for HumanManager {
    fn process_order(&self) {
        println!("Manager is processing logic...");
    }

    fn attend_meeting(&self) {
        println!("Manager is boring at the meeting...");
    }

    fn get_rest(&self) {
        println!("Manager is taking a break...");
    }

    fn swing_the_lead(&self) {
        println!("Manager is watching reels...");
    }
}

/// [`RobotPacker`] - робот
pub struct RobotPacker {
    pub model: String,
}

impl WarehouseWorker for RobotPacker {
    fn process_order(&self) {
        println!("Robot {} is packing boxes...", self.model);
    }

    fn attend_meeting(&self) {
        println!("ERROR: Robot cannot attend meetings");
    }

    fn get_rest(&self) {
        println!("Robot was taken for maintenance");
    }

    fn swing_the_lead(&self) {
        panic!("CRITICAL ERROR: Robot cannot waste our money (we hope so)");
    }
}

/// [`manage_warehouse`] - функция, которая работает со списком работников
pub fn manage_warehouse(workers: &[Box<dyn WarehouseWorker>]) {
    println!("\n--- Warehouse Shift Started ---");
    for worker in workers {
        worker.process_order();
        worker.attend_meeting();
        worker.get_rest();
        worker.swing_the_lead();
    }
}
